### ISSUE TITLE
Map source elements to a specific instance in a target array

### ISSUE THREAD

**Issue by vratojr** [atlasmap/atlasmap#2007]
Hello, 
I'd like to develop the following feature.

A screenshot is worth 1000 words so here it is :) :
![image](https://user-images.githubusercontent.com/1368836/81070982-56501d00-8ee4-11ea-889f-95d608369615.png)
The idea is that of being able to map to specific instances in an array. So the idea is that of adding a + sign on each array to add a specific instance over which the mapping can be defined. And a minus to remove it.

Would you be interested in integrating this new feature? 
If yes could you provide some guidance concerning how you would implement it? And if you are ok concerning the ui (well, we'd make it a bit better of course :) ).

PS: I know the title is almost the same to issue #1418 ... Indeed it's a kind of superset of that issue (maybe you could rename 1418 to "Map multiple source elements to their K/V instances in a target array"?)

**Comment by zamrokk** [comment]
It is similar to what is doing IBM WTX : https://www.youtube.com/watch?v=uVxYPH2gwM4
I vote for it !

**Comment by igarashitm** [comment]
+1 This is a nice to have.
I would enhance `Mapping Details` pane to handle this kind of usecase, instead of showing it in document itself. Internally we can add a transformation to put a value into specific indexed collection item, like `putAt()`.

**Comment by igarashitm** [comment]
By the way, please note that we're migrating to React version of UI and Angular version will be removed in 2.0.

**Comment by zamrokk** [comment]
If you keep the mapping of items on the panel then it is not very nice to do drag n drop to the item directly on the tree. I would prefer like WTX does

**Comment by zamrokk** [comment]
Btw, why react and no more angular?

**Comment by zamrokk** [comment]
Also you do have usage of right click to do actions on an item of the tree. Is it a choice to? Do you prefer contextual buttons on a tree?

**Comment by riccardo-forina** [comment]
We don't have any contextual menu or action in order to support assistive technologies and touch devices. In the React version we have a kind of toolbar that is shown on top of the field for mouse users with all the actions that can be performed on the field, which is also available to keyboard users moving through the app using tabs and arrows, and the enter key to focus a field.

![2020-05-06 at 16 24 53 - Emerald Mammal](https://user-images.githubusercontent.com/966316/81188954-486acc80-8fb6-11ea-94cc-7e2aa5487d6a.gif)

Atlasmap moved to React in order to use [PatternFly](https://www.patternfly.org/v4/) as the UI library.

**Comment by igarashitm** [comment]
We should also consider how to handle merging multiple collection, but not only mapping individual collection item. And also nested collection like `Company<>/Section<>/Employee<>` has to be taken into account.

**Comment by vratojr** [comment]
> We should also consider how to handle merging multiple collection, but not only mapping individual collection item. And also nested collection like `Company<>/Section<>/Employee<>` has to be taken into account.

@igarashitm Could you please expand a bit more the first problem?
We are interesting in also solving the problem of nested collections but for us it's less important than having this feature first. Would it be possible to tackle this problem in a second moment?

**Comment by vratojr** [comment]
@riccardo-forina in this case we could use that contextual menu to add + and -, right?

**Comment by riccardo-forina** [comment]
> @riccardo-forina in this case we could use that contextual menu to add + and -, right?

I guess so, although I'm inclined to think that this kind of transformation needs to be specified in the mapping details as a second step as @igarashitm was suggesting.

**Comment by igarashitm** [comment]
@vratojr for example
`Company1:/Employee<>/id` + `Company2:/Employee<>/id` <-> `Report:/ParticipantIdList<>`
I think both direction have to be supported when we tackle this collection enhancement.

**Comment by vratojr** [comment]
Frankly I can't imagine a representation using the mapping details pane that would be clear and provide flexibility. Could you please share any drawing?

**Comment by vratojr** [comment]
@igarashitm I see, thanks. What about tackling with nested arrays in a second moment?

**Comment by igarashitm** [comment]
If we can make sure the 1st step won't prevent further step, that would be OK. The drawing - user interface design is the biggest part for this and I don't have a good idea ATM. It would require a separate UI component for sure, maybe drilling down from somewhere and open up a "Collection Mapping" pane that is a special version of Mapping Details, or like that.

**Comment by igarashitm** [comment]
And of course it shouldn't break currently supported usecase in any step.

**Comment by vratojr** [comment]
Ok, concerning the UI we have two problems to solve:

1) We have to display that there's a mapping between elements and the array itself. (and viceversa)
2) We have to display the detail of the mapping between the elements and the elements in each item of the array. (and viceversa)

Let's start with the easy situation:
We have input property P and an output Array A.
We want to map P to the property C of the first and second elements of A
We want to map P to the property D of the third element of A

For the first point the display will be the same as actual situation:
![image](https://user-images.githubusercontent.com/1368836/81692698-1696b180-945f-11ea-8682-520365ef232d.png)
This works also for the reverse case and array to array mapping.

Second point:
We could add a transformation: "Map to specific item" and ask as input the index of the item to map to. This would also let us to concatenate specific transformations for each object:
![image](https://user-images.githubusercontent.com/1368836/81689592-b9006600-945a-11ea-8934-ca3073fc10f2.png)
This, would potentially also work when mapping an array to a property by using the transformation "Map from specific item" and also with array to array when mixing both.

When using **nested arrays,** the point 1 is still valid. For the point 2 we should enrich the ui since there is ambiguity:
Let's say that A's elements are objects containing arrays with properties C and D. 
A<>/B<>/{C,D} (I hope the notation is coherent with yours :) )

The user could want to map the value of P to:
* property C of the nth element of B of  mth element of A or
* property C of all elements of B of the mth element of A or
* property C of nth element of B of the mth element of A

In that case this could be solved by adding a "Parent collection field" in the mapping and by adding subsequents transformations like:
![image](https://user-images.githubusercontent.com/1368836/81696641-7ba0d600-9464-11ea-8a31-0e7c8d1da222.png)

So here's the first proposition over which we can start to discuss. To me it seems pretty compact and reuses a lot of components that you've already created.

**Comment by vratojr** [comment]
@igarashitm and @riccardo-forina, friendly ping :)

**Comment by riccardo-forina** [comment]
Hey, sorry for the lack of communication, we were at the end of the sprint and had to focus on what was planned. Now we are in code freeze and this should give us the chance to look at this.

**Comment by igarashitm** [comment]
Took longer to get out from busy sprint - not yet sure if I understood correctly,  from the screenshots, you want to add "map to item" transformation in a regular transformation chain? doesn't it mix up with other transformations that applies to whole collection? I was thinking to isolate it to "Collection Mapping" that is an advanced version of "Mapping Details" dedicated to collection mapping.
This "how to let user edit mapping multiplicity" has been a biggest headache for me, and nested collection killed me 100 times...

**Comment by vratojr** [comment]
It mixes, yes, but only to the current mapping.  
I mean, it won't be possible to define a generic mapping:
`P->A<>`
An then overwrite it with a more specific one:
`P->A<1>`
But I was thinking that it could be reasonable to say:"if you decide to specify a mapping to a given collection element, then you have to do it for each each other element". That is, if you map to element 1 then you have also to specify the mapping to the other elments, otherwise the elements will be left empty. Anyway I think that this kind of mapping will be mainly used to map an object to an array and so the set of properties to map will be countable.

**Comment by igarashitm** [comment]
I mean, with that reason I don't like to have "Map to item" as one of transformation dropdown item.

**Comment by vratojr** [comment]
So what about adding a button "Collection Mapping" that open this Collection Mapping in which you can define the mappings in the way I showed above?
In this way you can define global mappins from the Mapping Details and specific ones from the Collection Mapping.

**Comment by igarashitm** [comment]
Yes that sounds better. With that direction "Map to item" no longer have to be in dropdown menu, it can focus on collection item with its index in that specialized collection mapping pane.

**Comment by riccardo-forina** [comment]
Seconded

**Comment by vratojr** [comment]
What about the implementation? We group all the actions from the collection
pane in a separate list?

Il mar 16 giu 2020, 19:38 Riccardo Forina <notifications@github.com> ha
scritto:

**Comment by vratojr** [comment]
After thinking a bit more, I think it's more clear like this:

![image](https://user-images.githubusercontent.com/1368836/84877967-7816f500-b089-11ea-83e1-303f68abeed7.png)

In which we use two sub panes. The content of the first is the same as what we have now. In the second we also add the two mapping actions.
Like it?

**Comment by igarashitm** [comment]
I like it. Only one concern is that the vertical space is relatively expensive on modern wide display. Another option is to add some button in between "Mapping Details" and "X" button. I guess from this point PatternFly guildeline sometimes restricts what we can do.

**Comment by vratojr** [comment]
I see your point.
If we try to add some button to the header like you are suggesting, I think it cannot be textual since that component will become too much cluttered up. 
On the other side, we could think about adding just a action icon, like:
![image](https://user-images.githubusercontent.com/1368836/84916107-eecde580-b0bd-11ea-8b70-b593837fa35c.png)
But I think that we should still find some place where to put the information that we are defining the default or element mapping. So it will be a kind of header somewhere and I think it will be very similar (maybe we'll spare some pixels) to the solution I've already proposed using the navs. Do you have any idea on the subject?

On the other side, instead of using horizontal navs, we could use a vertical one and put it on the side but I'm not very fond of it...

But, if we want to gain some pixel, we can look elsewhere ;) For example. Currently we have a lot of space that's reserved for the conditional mapping expression. We could change this if we move the f(x) in the upper toolbar and activate the expression field only on click:

off:
![image](https://user-images.githubusercontent.com/1368836/84917083-28ebb700-b0bf-11ea-8fad-fe0713df6627.png)

on:
![image](https://user-images.githubusercontent.com/1368836/84917205-5173b100-b0bf-11ea-8beb-bd493a28d89e.png)

The entire application would profit from this small change.

**Comment by gashcrumb** [comment]
I think it's worth exploring.  We'd need to indicate to the user somehow on that icon if there's an expression in place if it's hidden.  If this approach was followed it'd be better probably if when expanded it floated above the content like a panel rather than make the UI shift downwards when expanded.  Kind of like the [popover](https://www.patternfly.org/v4/documentation/react/components/popover) perhaps.  That being said, the control itself has it's own typeahead/suggestion elements and I'd be concerned this would further complicate the usability of this field.

Follow on question, @pleacu or @riccardo-forina the conditional mapping expression field has to do with the currently selected mapping, right?  Do you recall the reason why it's in an area outside of the mapping details screen?

**Comment by igarashitm** [comment]
> the conditional mapping expression field has to do with the currently selected mapping, right? Do you recall the reason why it's in an area outside of the mapping details screen?

I can answer for this one, Mapping Details pane is a vertical space and we thought it's horizontally too short to write down a expression. We imagined spreadsheet math formula when we added Conditional Mapping (expression).

**Comment by riccardo-forina** [comment]
Semantically it makes sense in the details pane tho. The horizontal space limit can be overcome by making the textarea autoexpand in its height as needed (up to a limit). A bit like how Excel handles this.

About D&D, we talked during the last UX sync meeting and we felt that since the input method is already the keyboard, using drag & drop doesn't feel like the first input choice from a user perspective and so it's maybe not that important to support.

I think it makes a lot of sense to at least try moving it in the details pane and see how it feels.

**Comment by igarashitm** [comment]
Not a bad idea focusing on keyboard experience - I would even disable d&d for expression if we move it to the right side.
To expanding text field, if it's in Mapping Details pane, it should expand to left, going over targets and even sources I guess to get more horizontal space. It could be a full screen expression editor.

**Comment by gashcrumb** [comment]
>To expanding text field, if it's in Mapping Details pane, it should expand to left, going over targets and even sources I guess to get more horizontal space. It could be a full screen expression editor.

Kind of along [these lines](https://www.patternfly.org/v4/documentation/react/demos/masterdetaildemo/master-detail-full-page) perhaps?  I wonder how wide they let that detail pane expand.

**Comment by riccardo-forina** [comment]
The drawer is fixed to 500px I believe, or fullscreen on small screens.

**Comment by vratojr** [comment]
What about moving this discussion to a separate issue/pr so here we can focus on collection mapping? :)

**Comment by vratojr** [comment]
So, assuming that we are going to spare some vertical space, do we agree on the solution with tabs that I propose previously? Or the one with the action or other? ;)

**Comment by igarashitm** [comment]
Maybe go with tab at this point and re-evaluate the look&feel as a part of wider discussion for "Variation of Mapping Details pane"?

**Comment by vratojr** [comment]
Ok, concerning implementation. When we send the actions to the backend, we wrap then into a separate array 
`collectionAction`
for the given field?

For example,
````
actions: [],
collectionActions: [],
docId: "input_nested"
fieldType: "STRING"
index: 0
jsonType: "io.atlasmap.json.v2.JsonField"
name: "name"
path: "/A<>/B<>/name"
userCreated: false
````

so that we can first apply the 'global' actions and then overwrite them with the collection action?
Or do you prefer to add a flag to the Action object?

BTW: In the PR that I opened I have already implemented a stub of behavior for the backend. Could you check it and tell me if I'm on the good path?

**Comment by vratojr** [comment]
what about if we have multiple element mapping for the same field?
Say: when I map to element 0 do this, element 1 do another ect.
I cannot just modify the path. On top of this I have to trace all the other actions that should be applied to the give mapping (append, ect) so I need a list of actions.

**Comment by igarashitm** [comment]
Isn't it just 2 mappings?
`/foo -> /A[0]`
`/bar -> /A[1]`

**Comment by vratojr** [comment]
I thought about letting the user do this:

```
/foo -> A[0]
/foo -> A[1] and Prepend the string "$"
```

**Comment by igarashitm** [comment]
Then that prepend action could be applied to the 2nd mapping in "Collection Mapping" pane just like you do for non collection I guess?

**Comment by vratojr** [comment]
Yes but imagine:

```
/foo -> A[0]
/foo -> A[1] and Prepend the string "$"
/foo -> A[2] and do something else
```
:)
btw if you have a slack that we can use to communicate maybe it could go faster :)

**Comment by igarashitm** [comment]
BTW/FYI, we also have a gitter channel https://gitter.im/atlasmap/community for temporary discussion

**Comment by igarashitm** [comment]
Addressed by #2171
